File: sdks/ts/packages/golem-ts-sdk/src/fluent/keyvalue.ts
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Plain-async fluent wrapper around the `wasi:keyvalue@0.1.0` host interface.
// Every operation returns a `Promise` (or a plain value where the host call is
// synchronous) and throws a typed `KeyValueError` on failure.
//
// Only the `eventual` (single-key CRUD) and `eventual-batch` (multi-key CRUD)
// interfaces are exposed. The `atomic` and `cache` interfaces are NOT wrapped:
// they are `unimplemented!` in the Golem host and trap the worker.

import * as KvBatch from 'wasi:keyvalue/eventual-batch@0.1.0';
import { strictTextDecoder } from './strictTextDecoder';
import * as KvEventual from 'wasi:keyvalue/eventual@0.1.0';
import * as KvTypes from 'wasi:keyvalue/types@0.1.0';
import { compileSchema } from './schema/adapter';
import type { StandardSchemaV1 } from './schema/standardSchema';

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

const traceOf = (e: unknown): string => {
  if (e !== null && typeof e === 'object') {
    const obj = e as { trace?: () => string; message?: string };
    if (typeof obj.trace === 'function') {
      try {
        return obj.trace();
      } catch {
        /* fall through */
      }
    }
    if (typeof obj.message === 'string') return obj.message;
  }
  return String(e);
};

/**
 * Raised when any host call into `wasi:keyvalue@0.1.0` traps. The `trace`
 * field carries the verbatim driver-supplied string from the host's
 * `error.trace()` (or, for non-Error host throws, the stringified cause). Per
 * the Golem source this is opaque / driver-specific and should not be parsed.
 */
export class KeyValueError extends Error {
  override readonly name = 'KeyValueError';
  readonly trace: string;
  readonly operation: string;
  constructor(
    readonly cause: unknown,
    operation: string,
  ) {
    const t = traceOf(cause);
    super(`KeyValueError(${operation}): ${t}`);
    this.trace = t;
    this.operation = operation;
  }
}

const wrap = <A>(operation: string, fn: () => A): A => {
  try {
    return fn();
  } catch (cause) {
    if (cause instanceof KeyValueError) throw cause;
    throw new KeyValueError(cause, operation);
  }
};

// ---------------------------------------------------------------------------
// Schema codec — UTF-8 JSON over a Standard Schema
// ---------------------------------------------------------------------------
//
// The fluent SDK has no Effect Schema; we validate values through the Standard
// Schema's synchronous `~standard.validate` and JSON-encode the validated
// value to UTF-8 bytes. Validation/JSON failures throw {@link KeyValueError}.

const textEncoder = new TextEncoder();
const textDecoder = strictTextDecoder();

const validateSync = <T>(schema: StandardSchemaV1, value: unknown): T => {
  const result = schema['~standard'].validate(value);
  if (result instanceof Promise) {
    throw new Error('Schema validation must be synchronous for keyvalue payloads');
  }
  if (result.issues !== undefined) {
    const msg = result.issues.map((i) => i.message).join('; ');
    throw new Error(`Schema validation failed: ${msg}`);
  }
  return result.value as T;
};

/** Encode a schema-validated value to UTF-8 JSON bytes. */
const encodeValue = <T>(schema: StandardSchemaV1, value: T): Uint8Array =>
  wrap('schema.encode', () => {
    const validated = validateSync<T>(schema, value);
    return textEncoder.encode(JSON.stringify(validated));
  });

/** Decode UTF-8 JSON bytes and re-validate against the schema. */
const decodeValue = <T>(schema: StandardSchemaV1, bytes: Uint8Array): T =>
  wrap('schema.decode', () => {
    const json = textDecoder.decode(bytes);
    const parsed: unknown = JSON.parse(json);
    return validateSync<T>(schema, parsed);
  });

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/**
 * Schema-typed view of a {@link Bucket}. Values are validated against the
 * Standard Schema and JSON-encoded (UTF-8) on writes; reads JSON-parse and
 * re-validate. Decoding failures throw {@link KeyValueError}.
 */
export interface SchemaBucket<T> {
  get(key: string): T | undefined;
  set(key: string, value: T): void;
  delete(key: string): void;
  exists(key: string): boolean;
  getMany(keys: readonly string[]): Array<T | undefined>;
  setMany(entries: ReadonlyArray<readonly [string, T]>): void;
  deleteMany(keys: readonly string[]): void;
  keys(): string[];
}

/**
 * Handle on an open keyvalue bucket. Acquired via {@link openBucket}. The
 * underlying WIT `bucket` resource has no explicit close method; dropping the
 * JS handle is enough (GC-managed lifetime).
 *
 * All host calls are synchronous, so these methods return plain values rather
 * than Promises.
 */
export interface Bucket {
  readonly name: string;

  /** Get the bytes stored under `key`, or `undefined` if absent. */
  get(key: string): Uint8Array | undefined;
  /** Overwrite `key` with `value`, or insert if absent. */
  set(key: string, value: Uint8Array): void;
  /** Remove `key`. No-op if absent. */
  delete(key: string): void;
  /** Test whether `key` is present. */
  exists(key: string): boolean;

  /**
   * Batch-get. Returned array length === `keys.length`; missing keys surface
   * as `undefined`, positionally aligned with the request. All-or-nothing: a
   * storage-layer error fails the whole call.
   */
  getMany(keys: readonly string[]): Array<Uint8Array | undefined>;
  /** Batch-set. Per-entry order is not guaranteed; failure is partial. */
  setMany(entries: ReadonlyArray<readonly [string, Uint8Array]>): void;
  /** Batch-delete. Per-entry order is not guaranteed; failure is partial. */
  deleteMany(keys: readonly string[]): void;

  /** All keys currently present in the bucket. Order undefined. */
  keys(): string[];

  /**
   * Build a schema-typed view of this bucket. Values are JSON-encoded via the
   * supplied Standard Schema on writes and re-validated on reads.
   */
  forSchema<S extends StandardSchemaV1>(
    schema: S,
  ): SchemaBucket<StandardSchemaV1.InferOutput<S>>;
}

// ---------------------------------------------------------------------------
// Host plumbing helpers (IncomingValue / OutgoingValue chunking folded in)
// ---------------------------------------------------------------------------

const writeOutgoing = (bytes: Uint8Array): KvTypes.OutgoingValue =>
  wrap('outgoingValueWriteBodySync', () => {
    const ov = KvTypes.OutgoingValue.newOutgoingValue();
    ov.outgoingValueWriteBodySync(bytes);
    return ov;
  });

const consumeIncoming = (iv: KvTypes.IncomingValue): Uint8Array =>
  wrap('incomingValueConsumeSync', () => iv.incomingValueConsumeSync());

// ---------------------------------------------------------------------------
// SchemaBucket factory
// ---------------------------------------------------------------------------

const makeSchemaBucket = <T>(bucket: Bucket, schema: StandardSchemaV1): SchemaBucket<T> => ({
  get(key) {
    const raw = bucket.get(key);
    return raw === undefined ? undefined : decodeValue<T>(schema, raw);
  },
  set(key, value) {
    bucket.set(key, encodeValue(schema, value));
  },
  delete(key) {
    bucket.delete(key);
  },
  exists(key) {
    return bucket.exists(key);
  },
  getMany(keys) {
    return bucket
      .getMany(keys)
      .map((raw) => (raw === undefined ? undefined : decodeValue<T>(schema, raw)));
  },
  setMany(entries) {
    bucket.setMany(entries.map(([k, v]) => [k, encodeValue(schema, v)] as const));
  },
  deleteMany(keys) {
    bucket.deleteMany(keys);
  },
  keys() {
    return bucket.keys();
  },
});

// ---------------------------------------------------------------------------
// Bucket factory
// ---------------------------------------------------------------------------

const makeBucket = (name: string, handle: KvTypes.Bucket): Bucket => {
  const self: Bucket = {
    name,
    get(key) {
      const iv = wrap('eventual.get', () => KvEventual.get(handle, key));
      if (iv === undefined || iv === null) return undefined;
      return consumeIncoming(iv);
    },
    set(key, value) {
      const ov = writeOutgoing(value);
      wrap('eventual.set', () => KvEventual.set(handle, key, ov));
    },
    delete(key) {
      wrap('eventual.delete', () => KvEventual.delete_(handle, key));
    },
    exists(key) {
      return wrap('eventual.exists', () => KvEventual.exists(handle, key));
    },
    getMany(keys) {
      const raw = wrap(
        'eventual-batch.get-many',
        () =>
          KvBatch.getMany(handle, [...keys]) as ReadonlyArray<KvTypes.IncomingValue | undefined>,
      );
      return raw.map((iv) => (iv === undefined || iv === null ? undefined : consumeIncoming(iv)));
    },
    setMany(entries) {
      const pairs: Array<[string, KvTypes.OutgoingValue]> = entries.map(([k, v]) => [
        k,
        writeOutgoing(v),
      ]);
      wrap('eventual-batch.set-many', () => KvBatch.setMany(handle, pairs));
    },
    deleteMany(keys) {
      wrap('eventual-batch.delete-many', () => KvBatch.deleteMany(handle, [...keys]));
    },
    keys() {
      return wrap('eventual-batch.keys', () => KvBatch.keys(handle));
    },
    forSchema<S extends StandardSchemaV1>(schema: S) {
      // compileSchema validates the schema is a real, registered Standard
      // Schema (throwing a descriptive error otherwise) before we build the
      // JSON byte view over it.
      compileSchema(schema);
      return makeSchemaBucket<StandardSchemaV1.InferOutput<S>>(self, schema);
    },
  };
  return self;
};

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Open a bucket by name. The WIT bucket resource has no explicit close; the
 * returned handle's lifetime is GC-managed. Failures from the host's
 * `bucket.open-bucket` surface as {@link KeyValueError}.
 *
 * `openBucket` is async for API ergonomics, although the host call is
 * synchronous.
 */
export async function openBucket(name: string): Promise<Bucket> {
  const handle = wrap('openBucket', () => KvTypes.Bucket.openBucket(name));
  return makeBucket(name, handle);
}

File: sdks/go/golem/keyvalue/keyvalue.go
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package keyvalue is a Go wrapper over Golem's durable key-value store
// (wasi:keyvalue). Open a [Bucket] by name and Get/Set/Delete/Exists byte values,
// or wrap it with [Typed] for a JSON-encoded [Store] of a Go type.
//
// Unlike the exactly-once control-flow surface (RPC, promises, durability,
// transactions), which is fail-loud, these are I/O operations whose failures a
// caller can handle, so they return an error. A missing key is not an error: Get
// returns (nil, false, nil). The store is durable — operations are journaled and
// replayed — but because they are remote side effects, calling them inside a
// read-only method traps.
//
// Pair a fallible call with golem.Must / golem.Must0 / golem.Must2 to abort the
// invocation on error.
package keyvalue

import (
	"encoding/json"
	"errors"
	"fmt"

	eventual "github.com/golemcloud/golem/sdks/go/golem/internal/wit/wasi_keyvalue_eventual"
	batch "github.com/golemcloud/golem/sdks/go/golem/internal/wit/wasi_keyvalue_eventual_batch"
	kvtypes "github.com/golemcloud/golem/sdks/go/golem/internal/wit/wasi_keyvalue_types"
	kverr "github.com/golemcloud/golem/sdks/go/golem/internal/wit/wasi_keyvalue_wasi_keyvalue_error"
	witTypes "go.bytecodealliance.org/pkg/wit/types"
)

// Error is a key-value host error, carrying the host's diagnostic trace.
type Error struct{ Trace string }

func (e *Error) Error() string { return "golem/keyvalue: " + e.Trace }

func kvError(e *kverr.Error) error {
	if e == nil {
		return errors.New("golem/keyvalue: unknown error")
	}
	return &Error{Trace: e.Trace()}
}

// Bucket is a handle to a key-value bucket.
type Bucket struct{ raw *kvtypes.Bucket }

// OpenBucket opens (or creates) the named bucket.
func OpenBucket(name string) (*Bucket, error) {
	r := kvtypes.BucketOpenBucket(name)
	if r.IsErr() {
		return nil, kvError(r.Err())
	}
	return &Bucket{raw: r.Ok()}, nil
}

// Get returns the value stored at key. found is false (with a nil error) when the
// key is absent.
func (b *Bucket) Get(key string) (value []byte, found bool, err error) {
	r := eventual.Get(b.raw, key)
	if r.IsErr() {
		return nil, false, kvError(r.Err())
	}
	opt := r.Ok()
	if opt.IsNone() {
		return nil, false, nil
	}
	return consume(opt.Some())
}

// Set stores value at key.
func (b *Bucket) Set(key string, value []byte) error {
	ov, err := outgoing(value)
	if err != nil {
		return err
	}
	if r := eventual.Set(b.raw, key, ov); r.IsErr() {
		return kvError(r.Err())
	}
	return nil
}

// Delete removes key (a no-op if it is absent).
func (b *Bucket) Delete(key string) error {
	if r := eventual.Delete(b.raw, key); r.IsErr() {
		return kvError(r.Err())
	}
	return nil
}

// Exists reports whether key is present.
func (b *Bucket) Exists(key string) (bool, error) {
	r := eventual.Exists(b.raw, key)
	if r.IsErr() {
		return false, kvError(r.Err())
	}
	return r.Ok(), nil
}

// Keys lists all keys in the bucket.
func (b *Bucket) Keys() ([]string, error) {
	r := batch.Keys(b.raw)
	if r.IsErr() {
		return nil, kvError(r.Err())
	}
	return r.Ok(), nil
}

// GetMany fetches several keys at once, returning a map of only the present keys.
// The batch is not atomic.
func (b *Bucket) GetMany(keys []string) (map[string][]byte, error) {
	r := batch.GetMany(b.raw, keys)
	if r.IsErr() {
		return nil, kvError(r.Err())
	}
	opts := r.Ok()
	out := make(map[string][]byte, len(opts))
	for i, o := range opts {
		if i >= len(keys) || o.IsNone() {
			continue
		}
		v, _, err := consume(o.Some())
		if err != nil {
			return nil, err
		}
		out[keys[i]] = v
	}
	return out, nil
}

// SetMany stores several entries at once. The batch is not atomic.
func (b *Bucket) SetMany(entries map[string][]byte) error {
	kvs := make([]witTypes.Tuple2[string, *kvtypes.OutgoingValue], 0, len(entries))
	for k, v := range entries {
		ov, err := outgoing(v)
		if err != nil {
			return err
		}
		kvs = append(kvs, witTypes.Tuple2[string, *kvtypes.OutgoingValue]{F0: k, F1: ov})
	}
	if r := batch.SetMany(b.raw, kvs); r.IsErr() {
		return kvError(r.Err())
	}
	return nil
}

// DeleteMany removes several keys at once. The batch is not atomic.
func (b *Bucket) DeleteMany(keys []string) error {
	if r := batch.DeleteMany(b.raw, keys); r.IsErr() {
		return kvError(r.Err())
	}
	return nil
}

func consume(iv *kvtypes.IncomingValue) ([]byte, bool, error) {
	r := iv.IncomingValueConsumeSync()
	if r.IsErr() {
		return nil, false, kvError(r.Err())
	}
	return r.Ok(), true, nil
}

func outgoing(value []byte) (*kvtypes.OutgoingValue, error) {
	ov := kvtypes.OutgoingValueNewOutgoingValue()
	if r := ov.OutgoingValueWriteBodySync(value); r.IsErr() {
		return nil, kvError(r.Err())
	}
	return ov, nil
}

// ── Typed store ───────────────────────────────────────────────────────────────

// Store is a typed view over a [Bucket]: values are JSON-encoded T. Build one with
// [Typed].
type Store[T any] struct{ b *Bucket }

// Typed wraps a bucket as a JSON-encoded store of T.
func Typed[T any](b *Bucket) *Store[T] { return &Store[T]{b: b} }

// Get decodes the value at key as T. found is false (nil error) when absent.
func (s *Store[T]) Get(key string) (value T, found bool, err error) {
	raw, found, err := s.b.Get(key)
	if err != nil || !found {
		var zero T
		return zero, found, err
	}
	v, err := unmarshalValue[T](raw)
	if err != nil {
		var zero T
		return zero, true, fmt.Errorf("golem/keyvalue: decoding %q: %w", key, err)
	}
	return v, true, nil
}

// Set JSON-encodes value and stores it at key.
func (s *Store[T]) Set(key string, value T) error {
	raw, err := marshalValue(value)
	if err != nil {
		return fmt.Errorf("golem/keyvalue: encoding %q: %w", key, err)
	}
	return s.b.Set(key, raw)
}

// Delete removes key.
func (s *Store[T]) Delete(key string) error { return s.b.Delete(key) }

// Exists reports whether key is present.
func (s *Store[T]) Exists(key string) (bool, error) { return s.b.Exists(key) }

// Keys lists all keys.
func (s *Store[T]) Keys() ([]string, error) { return s.b.Keys() }

// marshalValue/unmarshalValue are the pure JSON codec used by Store (no host
// calls), so they are natively testable.
func marshalValue[T any](v T) ([]byte, error) { return json.Marshal(v) }
func unmarshalValue[T any](data []byte) (T, error) {
	var v T
	err := json.Unmarshal(data, &v)
	return v, err
}
